juju/model.py

```diff
@@ -0,0 +1,166 @@
+        # TODO Do we want to move to UUIDs? Is that even possible with the number of
+        # places that rely on unit names being of the form {app_name}/{unit_number}?
+        # Note that CMRs don't expose the actual unit name, but still follow the form,
+        # using someething like remote-56a0f163eb4e4f2e88d50983dca7be02/0.
+        self.name = name
+
+    def __hash__(self):
+        # A Unit instance from one relation will not be identical to one from another relation.
+        # However, they should be considered equivalent, since they are views of the same entity.
+        # Generally, this should only matter for using model.unit as an index into a given relation's
+        # relation data.
+        return hash(self.name)
+
+    def __eq__(self, other):
+        if not isinstance(other, Unit):
+            raise TypeError(f"'==' not supported between instances of '{type(self).__name__}' and '{type(other).__name__}'")
+        return self.name == other.name
+
+
+class RelationMap(dict):
+    """Map of relation names to lists of Relation instances.
+
+    The data for each relation name may be lazy-loaded.
+    """
+    def __init__(self, data):
+        super().__init__()
+        for relation_name, relations in data.items():
+            if relations is not None:
+                relations = [Relation(relation_id, relation_data)
+                             for relation_id, relation_data in sorted(relations.items())]
+            self[relation_name] = relations
+
+    def __getitem__(self, relation_name):
+        value = super().__getitem__(relation_name)
+        if value is None:
+            # Lazy-load the list of relations for this relation name.
+            output = run(['relation-ids', relation_name, '--format=yaml'], stdout=PIPE, check=True).stdout
+            value = [Relation(relation_id) for relation_id in yaml.safe_load(output.decode('utf8'))]
+            self[relation_name] = value
+        return value
+
+
+class Relation:
+    def __init__(self, relation_id, data=None):
+        self.relation_id = relation_id
+        self.data = RelationData(relation_id, data)
+        if data is not None:
+            self._apps = [key for key in self.data if isinstance(key, Application)]
+        else:
+            self._apps = None
+
+    @property
+    def apps(self):
+        if self._apps is None:
+            # Lazy-load the apps on this relation.
+            output = run(['relation-list', '-r', self.relation_id, '--format=yaml'], stdout=PIPE, check=True).stdout
+            member_names = yaml.safe_load(output.decode('utf8'))
+            units_by_app = {}
+            for member_name in sorted(member_names):
+                if '/' not in member_name:
+                    # Handle possible future support from Juju for application relation data.
+                    continue
+                app_name = member_name.split('/')[0]
+                units_by_app.setdefault(app_name, []).append(Unit(member_name))
```

Comment:
This is creating a list on every iteration just to throw it away in case it's not found, and it will be found at all times but the first one. This is also not a case where we are at a trivial path anymore, in the sense this is a for loop inside other for loops already. So it's worth paying attention to those trivial optimizations which don't really cost anything in terms of coding other than just reshaping the code easily.

I wish that was a non-issue because Python would optimize the code away.. but AFAIK it doesn't still.